### PR TITLE
fix bug when parse url start with http://

### PR DESCRIPTION
**Pull request by imaginezz**
When a markdown document have links which start with "http://", then will get Invalid URL error in `hexo generate` or `hexo server` command, such as:

```
ERROR Process failed: _posts/test.md
TypeError [ERR_INVALID_URL]: Invalid URL: http://
    at onParseError (internal/url.js:257:9)
    at new URL (internal/url.js:333:5)
    at new URL (internal/url.js:330:22)
    at /Users/jason/Documents/blog_store/node_modules/hexo-util/lib/full_url_for.js:22:18
    at Cache.apply (/Users/jason/Documents/blog_store/node_modules/hexo-util/lib/cache.js:27:46)
    at Hexo.fullUrlForHelper (/Users/jason/Documents/blog_store/node_modules/hexo-util/lib/full_url_for.js:18:16)
    at Hexo.module.exports (/Users/jason/Documents/blog_store/node_modules/hexo/lib/plugins/helper/full_url_for.js:5:23)
    at _Document.<anonymous> (/Users/jason/Documents/blog_store/node_modules/hexo/lib/models/post.js:54:25)
    at _Document.permalink (/Users/jason/Documents/blog_store/node_modules/warehouse/lib/types/virtual.js:59:24)
    at _Document.toObject (/Users/jason/Documents/blog_store/node_modules/warehouse/lib/document.js:73:44)
    at _Model._insertOne (/Users/jason/Documents/blog_store/node_modules/warehouse/lib/model.js:157:25)
    at /Users/jason/Documents/blog_store/node_modules/warehouse/lib/model.js:179:63
    at tryCatcher (/Users/jason/Documents/blog_store/node_modules/bluebird/js/release/util.js:16:23)
    at /Users/jason/Documents/blog_store/node_modules/bluebird/js/release/using.js:185:26
    at tryCatcher (/Users/jason/Documents/blog_store/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/Users/jason/Documents/blog_store/node_modules/bluebird/js/release/promise.js:547:31)
    at Promise._settlePromise (/Users/jason/Documents/blog_store/node_modules/bluebird/js/release/promise.js:604:18)
    at Promise._settlePromise0 (/Users/jason/Documents/blog_store/node_modules/bluebird/js/release/promise.js:649:10)
    at Promise._settlePromises (/Users/jason/Documents/blog_store/node_modules/bluebird/js/release/promise.js:729:18)
    at Promise._fulfill (/Users/jason/Documents/blog_store/node_modules/bluebird/js/release/promise.js:673:18)
    at PromiseArray._resolve (/Users/jason/Documents/blog_store/node_modules/bluebird/js/release/promise_array.js:127:19)
    at PromiseArray._promiseFulfilled (/Users/jason/Documents/blog_store/node_modules/bluebird/js/release/promise_array.js:145:14)
```

### CODE REVIEW

**Comment by SukkaW** [comment]
Please adding a new test case for your changes.

**Comment by SukkaW** [comment]
@imaginezz Please stop mentioning your PR in not related issue.
If you can not provide any test cases for your changes, your PR will be closed.

**Comment by imaginezz** [comment]
@SukkaW Sorry, I don't know the rule, I just think thir pr probably can help solve that issue. And I add the test case for my pr in a new commit.

File: lib/full_url_for.js [hunk 1]
```diff
@@ -19,10 +19,12 @@ function fullUrlForHelper(path = '/') {
...
-    const data = new URL(path, `http://${sitehost}`);
-
-    // Exit if input is an external link or a data url
-    if (data.hostname !== sitehost || data.origin === 'null') return path;
+    let data = null;
+    try {
+      data = new URL(path, `http://${sitehost}`);
+      // Exit if input is an external link or a data url
+      if (data.hostname !== sitehost || data.origin === 'null') return path;
+    } catch {}
```
Comment (SukkaW) [inline lib/full_url_for.js:27]:
```suggestion
    let data;
    try {
      data = new URL(path, `http://${sitehost}`);
    } catch {}

    // if there is invalid url, data should be undefined
    if (typeof data !== 'object') return path;

    // Exit if input is an external link or a data url
    if (data.hostname !== sitehost || data.origin === 'null') return path;
```

- Use `typeof data !== 'object'` as it has consistent behavior, see https://github.com/hexojs/hexo-util/blob/d08cd68582a39cb0f3357f5d038a6b266551cfc9/lib/is_external_link.js#L31-L35
- `try..catch` without error parameter syntax is unstable, please avoid using this.

Comment (imaginezz) [reply lib/full_url_for.js]: I've refactored the code, please review again.

**Comment by imaginezz** [comment]
I've refactored the code, please review again.

File: lib/full_url_for.js [hunk 2]
```diff
@@ -18,11 +18,14 @@ function fullUrlForHelper(path = '/') {
...
     if (/^(\/\/|http(s)?:)/.test(path)) return path;
 
-    const sitehost = parse(config.url).hostname || config.url;
-    const data = new URL(path, `http://${sitehost}`);
-
-    // Exit if input is an external link or a data url
-    if (data.hostname !== sitehost || data.origin === 'null') return path;
+    const config_url = config.url;
+    // if sitehost is empty, skip this step
+    if (config_url !== '' && config_url !== null) {
```
Comment (SukkaW) [inline lib/full_url_for.js:23]: It is not safe, in case no `config.url` is given the `undefined` will be passed then.

Comment (imaginezz) [reply lib/full_url_for.js:24]: @SukkaW  You are right, and I add code and test cases for handling the null and undefined conditions.
